Core/src/com/infiniteautomation/mango/spring/service/ConfigurationTemplateService.java

```diff
@@ -0,0 +1,155 @@
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.core.io.ClassPathResource;
+import org.springframework.stereotype.Service;
+
+import com.github.mustachejava.DefaultMustacheFactory;
+import com.github.mustachejava.Mustache;
+import com.github.mustachejava.MustacheFactory;
+import com.serotonin.m2m2.Common;
+import com.serotonin.m2m2.vo.permission.PermissionException;
+
+import au.com.bytecode.opencsv.CSVReader;
+
+@Service
+public class ConfigurationTemplateService {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ConfigurationTemplateService.class);
+
+    private final FileStoreService fileStoreService;
+
+    private final PermissionService permissionService;
+
+    public ConfigurationTemplateService(FileStoreService fileStoreService, PermissionService permissionService) {
+        this.fileStoreService = fileStoreService;
+        this.permissionService = permissionService;
+    }
+
+    /**
+     *
+     * @param fileStore
+     * @param filePath
+     * @return
+     * @throws IOException
+     * @throws PermissionException
+     */
+    public List<Map<String, Object>> generateConfig(String fileStore, String filePath) throws IOException,
+            PermissionException {
+        permissionService.ensureAdminRole(Common.getUser());
+
+        Path path = fileStoreService.getPathForRead(fileStore, filePath);
+        try (Reader in = new InputStreamReader(Files.newInputStream(path), Charset.defaultCharset())) {
+            CSVReader csvReader = new CSVReader(in);
+            List<Map<String, Object>> readFileStructure = readCSV(csvReader);
+            List<Map<String, String>> keys = new ArrayList<>(){{
+                add(new HashMap<>(){{
+                    put("groupKey", "site");
+                    put("childrenKey", "dataHalls");}});
+                add(new HashMap<>(){{
+                    put("groupKey", "dataHall");
+                    put("childrenKey", "devices");}});
+            }};
+
+            List<Map<String, Object>> result = groupByKeys(readFileStructure, keys);
+            return result;
+        }
+    }
+    public String generateTemplate(String fileStore, String filePath, String template) throws IOException,
+            PermissionException {
+        MustacheFactory mf = new DefaultMustacheFactory();
+        List<Map<String, Object>> sites = generateConfig(fileStore, filePath);
+
+        ClassPathResource cp = new ClassPathResource(template);
+        Mustache m = mf.compile(cp.getPath());
+        StringWriter writer = new StringWriter();
+        m.execute(writer, sites).flush();
+        System.out.println("Testing result" + writer.toString());
+        return writer.toString();
+    }
+
+    public static List<Map<String, Object>> groupByKey(List<Map<String, Object>> array, Map<String, String> key) {
```

Comment:
Any specific reason these are public and static?

Reply:
none, changed to private methods.